File: Flywaight/Python/LineaLigera.py
from abc import abstractclassmethod


class ILineaLigera:
    @abstractclassmethod
    def getColor():
        pass
    @abstractclassmethod
    def dibujar():
        pass

class Linea(ILineaLigera):
    Color:str
    def __init__(self,colo:str) -> None:
        self.Color=colo
    @staticmethod
    def getColor(self):
        return self.Color
    @staticmethod
    def dibujar(self,col,fil):
        print("Dibujando línea de color ["+self.Color+"] en ["+str(col)+", "+str(fil)+"]")

class FabricasDeLineas:
    lineas=[]
    def __init__(self) -> None:
        self.lineas.clear()
    
    def getLine(self,color:str):
        for i in self.lineas:
            if i.getColor(i)==color:
                print("Línea de color [" + color + "] encontrada, la devolvemos")
                return i
        print("Creando una línea de color [" + color + "]")
        self.lineas.append(Linea(color))
        return self.lineas[len(self.lineas)-1]

File: Flywaight/cpp/LineaLigera.cpp
#include <bits/stdc++.h>

using namespace std;

//-----------------interfaz----------------
class IlineaLigera{
    private:
        /* data */
    public:
        virtual string getColor()=0;
        virtual void dibujar(int,int)=0;
};

//----------------clase concreta-----------

class Linea:public IlineaLigera{
    private:
        string Color;
    public:
        Linea(string c);
        virtual string getColor();
        virtual void dibujar(int,int);
};

Linea::Linea(string c){
    Color=c;
}

string Linea::getColor(){
    return this->Color;
}

void Linea::dibujar(int col, int fila){
    cout << "Dibujando línea de color ["<<this->Color<<"] en ["<<col<<", "<<fila<< "]" <<endl;
}